ocaml/typing/env.ml

```diff
@@ -3189,44 +3261,52 @@ and lookup_apply ~errors ~use ~loc lid0 env =
   check_apply ~path:f0_path ~comp:f0_comp args0
 
 and lookup_module ~errors ~use ~loc lid env =
-  match lid with
-  | Lident s ->
-      let path, data = lookup_ident_module Load ~errors ~use ~loc s env in
-      let md = Subst.Lazy.force_module_decl data.mda_declaration in
-      path, md
-  | Ldot(l, s) ->
-      let path, data = lookup_dot_module ~errors ~use ~loc l s env in
-      let md = Subst.Lazy.force_module_decl data.mda_declaration in
-      path, md
-  | Lapply _ as lid ->
-      let path_f, comp_f, path_arg = lookup_apply ~errors ~use ~loc lid env in
-      let md = md (modtype_of_functor_appl comp_f path_f path_arg) in
-      Papply(path_f, path_arg), md
+  let path, locks, md =
+    match lid with
+    | Lident s ->
+        let path, locks, data = lookup_ident_module Load ~errors ~use ~loc s env in
+        let md = Subst.Lazy.force_module_decl data.mda_declaration in
+        path, locks, md
+    | Ldot(l, s) ->
+        let path, locks, data = lookup_dot_module ~errors ~use ~loc l s env in
+        let md = Subst.Lazy.force_module_decl data.mda_declaration in
+        path, locks, md
+    | Lapply _ as lid ->
+        let path_f, comp_f, path_arg = lookup_apply ~errors ~use ~loc lid env in
+        let md = md (modtype_of_functor_appl comp_f path_f path_arg) in
+        Papply(path_f, path_arg), [], md
+  in
+  if use then begin
+    let mda_mode = Mode.Value.legacy |> Mode.Value.disallow_right in
+    let vmode = walk_locks ~errors ~loc ~env ~item:Module ~lid mda_mode None locks in
+    submode_legacy ~errors ~loc ~env ~item:Module ~lid vmode
```

Comment:
I think this should be done by the callers of this function. We don't yet know what the module is being used for, so we don't know what mode is needed.

Reply:
Similar to the comment below, I think for the scope of this PR, it's quick and safe to just say all modules are always legacy.

Reply:
I find it hard to think about the correctness of the current code. This check might be too strict for some of the uses. If we just pull it out to the right places then it will be clear.

Reply:
Do you have an example where this might be too strict?